src/languages/gcode.js

```diff
@@ -7,58 +7,116 @@
  */
 
 export default function(hljs) {
-  const GCODE_IDENT_RE = '[A-Z_][A-Z0-9_.]*';
-  const GCODE_CLOSE_RE = '%';
   const GCODE_KEYWORDS = {
-    $pattern: GCODE_IDENT_RE,
-    keyword: 'IF DO WHILE ENDWHILE CALL ENDIF SUB ENDSUB GOTO REPEAT ENDREPEAT '
-      + 'EQ LT GT NE GE LE OR XOR'
-  };
-  const GCODE_START = {
-    className: 'meta',
-    begin: '([O])([0-9]+)'
+    $pattern: /[A-Z]+|%/,
+    meta: [
+      // conditions
+      'THEN',
+      'ELSE',
+      'ENDIF',
+      'IF',
+
+      // controls
+      'GOTO',
+      'DO',
+      'WHILE',
+      'WH',
+      'END',
+      'CALL',
+
+      // scoping
+      'SUB',
+      'ENDSUB',
+
+      // comparisons
+      'EQ',
+      'NE',
+      'LT',
+      'GT',
+      'LE',
+      'GE',
+      'AND',
+      'OR',
+      'XOR',
+
+      // start/end of program
+      '%'
+    ],
+    built_in: [
+      'ATAN',
+      'ABS',
+      'ACOS',
+      'ASIN',
+      'COS',
+      'EXP',
+      'FIX',
+      'FUP',
+      'ROUND',
+      'LN',
+      'SIN',
+      'SQRT',
+      'TAN',
+      'EXISTS'
+    ]
   };
-  const NUMBER = hljs.inherit(hljs.C_NUMBER_MODE, { begin: '([-+]?((\\.\\d+)|(\\d+)(\\.\\d*)?))|' + hljs.C_NUMBER_RE });
+
+
+  const NUMBER = /[+-]?((\.\d+)|(\d+)(\.\d*)?)/;
+
   const GCODE_CODE = [
-    hljs.C_LINE_COMMENT_MODE,
-    hljs.C_BLOCK_COMMENT_MODE,
+    // comments
     hljs.COMMENT(/\(/, /\)/),
-    NUMBER,
-    hljs.inherit(hljs.APOS_STRING_MODE, { illegal: null }),
-    hljs.inherit(hljs.QUOTE_STRING_MODE, { illegal: null }),
+    hljs.COMMENT(/;/, /$/),
+    hljs.APOS_STRING_MODE,
+    hljs.QUOTE_STRING_MODE,
```

Comment:
So we're sure we want the illegal matching?

Reply:
Quoted strings do not appear in the spec at all, but to my knowledge they are common enough to be included here. All implementations I've seen during my research were strict about 1 line equalling 1 command, which IS in the spec.

So I think newlines in quoted strings are a good indicator that the parsed text is not gcode.